### DISCUSSION TITLE
Generic PLC I/O communication for data exchange  with pycomm3

### DISCUSSION THREAD

**Discussion by gakos23** [ottowayi/pycomm3#259] [Q&A]
Hello there i need your help...i am trying to establish a communication to send/receive data between a Berghoff PLC(adapter) and a python script(scanner).At the same time i capture the traffic in wireshark.At first with the generic_message i send a service request to register a session and after i send a get_attribute_single request three times for assembly object input output and config.After the success responses i send a forward_open request to start implicit I/O messaging but the response says that this service is not supported.Any ideas??
Here i paste the main program and the wireshark capture

![Screenshot from 2023-01-26 16-38-18](https://user-images.githubusercontent.com/123645890/214868072-01383211-8ce6-43d1-8560-f56b9869a53f.png)
![Screenshot from 2023-01-26 16-45-48](https://user-images.githubusercontent.com/123645890/214868270-dbc9ebaa-f1ca-43c6-868d-ecc763d3a1e8.png)

**Comment by ottowayi** [comment]
I think it's because the request also needs to be routed to the message router object, so that's why it's saying the service isn't supported.  You're also missing all of the service parameters, you need to send along a lot of connection configuration as the request data.  Instead of trying to build the message from scratch, you can call the `CIPDriver._forward_open` method manually or set your generic messages to `connected=True` and it will do it automatically.

**Reply by gakos23** [reply]
Ok so i used enip explorer to set up an I/O connection between plc and my pc and i used the connection data i collected from wireshark to fill the request data parameters in open_forward message.Now i have a implicit connection but i have only responses from plc with data(T-->O) but i dont have 0-->T messages.Am i missing something??=
![Screenshot from 2023-01-27 13-27-18](https://user-images.githubusercontent.com/123645890/215076652-955d3eda-3a5a-402d-bad1-fa97095d72c0.png)
![Screenshot from 2023-01-27 13-30-21](https://user-images.githubusercontent.com/123645890/215076688-eb888acb-f9ab-4900-93aa-b332b308dc55.png)

**Reply by ottowayi** [reply]
Ohh... I'm sorry, I misunderstood what you are trying to do.   pycomm3 supports class 3 connections, which are request/response, where as I/O or implicit connections are scheduled and continuous.  What you're trying is not impossible, but will require a lot of more work.  You'll have to change the socket logic to always be reading for messages, instead of now only after sending a request.  It's been awhile since I looked into I/O connections, but I think they may also be UDP on a different port than 44818.  This functionality is something I plan on adding in the (distant) future, I'm currently rewriting the protocol implementation from scratch in a way that would make creating the adapter/server and class 1 implementations easier.  Unfortunately, life has gotten a bit busier for me lately, so I don't have a timeline on any of it right now.

**Reply by gakos23** [reply]
Ok i will try to do it!!!Thank you very much!!!
